espnet/asr/pytorch_backend/train.py

```diff
@@ -0,0 +1,214 @@
+
+    :param Namespace args: The program arguments
+    """
+    from espnet.asr.pytorch_backend.asr_job import PlotAttentionJob
+    from espnet.asr.pytorch_backend.asr_job import TrainingJob
+    from espnet.asr.pytorch_backend.asr_job import ValidationJob
+    from espnet.asr.pytorch_backend.dataset import ASRDataset
+    from espnet.utils.deterministic_utils import set_deterministic_pytorch
+    from espnet.utils.training.job import JobRunner
+
+    if args.distributed:
+        logging.info('use GPU {} for training'.format(gpuid))
+        # For multiprocessing distributed training, rank needs to be the
+        # global rank among all the processes
+        args.rank = args.rank * args.ngpu + gpuid
+        torch.distributed.init_process_group(backend=args.dist_backend, init_method=args.dist_url,
+                                             world_size=args.world_size, rank=args.rank)
+
+    # check cuda availability
+    if not torch.cuda.is_available():
+        logging.warning('cuda is not available')
+
+    model = build_model(args)
+    subsampling_factor = model.subsample[0]
+
+    if args.distributed:
+        # For multiprocessing distributed, DistributedDataParallel constructor
+        # should always set the single device scope, otherwise,
+        # DistributedDataParallel will use all available devices.
+        logging.info('batch size is automatically increased (%d -> %d)' % (
+            args.batch_size, args.batch_size * args.world_size))
+        torch.cuda.set_device(gpuid)
+        model.cuda(gpuid)
+        # When using a single GPU per process and per
+        # DistributedDataParallel, we need to divide the batch size
+        # ourselves based on the total number of GPUs we have
+        model = torch.nn.parallel.DistributedDataParallel(model, device_ids=[gpuid])
+    elif args.ngpu > 1:
+        logging.info('batch size is automatically increased (%d -> %d)' % (
+            args.batch_size, args.batch_size * args.ngpu))
+        model = torch.nn.DataParallel(model, device_ids=list(range(args.ngpu)))
+        model.cuda()
+        args.batch_size *= args.ngpu
+
+    # set torch device
+    device = torch.device("cuda" if args.ngpu > 0 else "cpu")
+    model = model.to(device)
+
+    optimizer = build_optimizer(args, model)
+
+    # load datasets
+    train_dataset = ASRDataset(load_batchset(args.train_json, args),
+                               subsampling_factor, args.preprocess_conf)
+    valid_dataset = ASRDataset(load_batchset(args.valid_json, args),
+                               subsampling_factor, args.preprocess_conf)
+
+    if args.distributed:
+        sampler = torch.utils.data.distributed.DistributedSampler(train_dataset)
+    else:
+        sampler = None
+    # batch_size=1 because minibatch is already made in dataset
+    train_loader = torch.utils.data.DataLoader(
+        train_dataset, batch_size=1, shuffle=sampler, num_workers=args.n_iter_processes, pin_memory=True)
```

Comment:
I think it should be `sampler=sampler` rather than `shuffle=sampler`,  this typo would make every single job process whole datasets

Reply:
@Cescfangs, can you review https://github.com/espnet/espnet/tree/v.0.7.0/espnet2 if you're interested in chainer-free stuff? We are now implementing a new trainer.

Reply:
> @Cescfangs, can you review https://github.com/espnet/espnet/tree/v.0.7.0/espnet2 if you're interested in chainer-free stuff? We are now implementing a new trainer.

Wow! that's cool, I'd like to.
However, I'm not familiar with the `espnet2` stuff, I may need couple days to play around it.